src/ledger/LedgerStateSnapshot.cpp

```diff
@@ -0,0 +1,296 @@
+// Copyright 2024 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "ledger/LedgerStateSnapshot.h"
+#include "bucket/BucketListSnapshot.h"
+#include "bucket/BucketManager.h"
+#include "bucket/BucketSnapshotManager.h"
+#include "ledger/LedgerTxn.h"
+#include "main/Application.h"
+#include "transactions/TransactionFrame.h"
+#include "transactions/TransactionUtils.h"
+#include "xdr/Stellar-ledger.h"
+
+namespace stellar
+{
+
+LedgerEntryWrapper::LedgerEntryWrapper(ConstLedgerTxnEntry&& entry)
+{
+    mEntry.emplace<ConstLedgerTxnEntry>(std::move(entry));
+}
+
+LedgerEntryWrapper::LedgerEntryWrapper(LedgerTxnEntry&& entry)
+{
+    mEntry.emplace<LedgerTxnEntry>(std::move(entry));
+}
+
+LedgerEntryWrapper::LedgerEntryWrapper(std::shared_ptr<LedgerEntry> entry)
+{
+    mEntry.emplace<std::shared_ptr<LedgerEntry>>(entry);
+}
+
+LedgerEntry const&
+LedgerEntryWrapper::current() const
+{
+    if (mEntry.index() == 0)
+    {
+        return std::get<0>(mEntry).current();
+    }
+    else if (mEntry.index() == 1)
+    {
+        return std::get<1>(mEntry).current();
+    }
+    else
+    {
+        return *std::get<2>(mEntry);
+    }
+}
+
+LedgerEntry&
+LedgerEntryWrapper::currentToModify()
+{
+    if (mEntry.index() == 0)
+    {
+        return std::get<0>(mEntry).current();
+    }
+    else if (mEntry.index() == 1)
+    {
+        throw std::runtime_error(
+            "LedgerEntryWrapper::current() called on a const LedgerEntry");
+    }
+    else
+    {
+        return *std::get<2>(mEntry);
```

Comment:
This should also throw on the `shared_ptr` case. Unlike in the `LedgerTxnEntry` case, changes to the shared ptr will never get reflected back to the ledger, so there should never be a reason to "modify" a BucketList returned shared_ptr (now that I think about it, I should change the return type of BucketListDB to `std::shared_ptr<LedgerEntry const>`, but I'll save that for one of my PRs).

Reply:
Yeah this case is a bit weird - in loadgen, we don't actually modify the entry, we just use a copy of loaded entry to construct a new upgrade config. With LedgerTxn, we use a nested child ledgerTxn, which gets rolled back. With BucketList snapshots, we use a copy of the ledger entry. I wonder if the name is confusing, and should be named something like `currentForTesting`? I also added a comment clarifying how this function is used.

Reply:
Oh, actually, I think we don't need this function at all - I can just copy the entry in loadgen, since it's the only use case of `LedgerEntryWrapper::currentToModify`